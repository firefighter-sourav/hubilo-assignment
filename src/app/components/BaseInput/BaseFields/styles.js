import styled from 'styled-components'

import { Flex } from 'app/configs/commonStyles'
import { colors } from 'app/themes/colorPalette'
export const Wrapper = styled.div`
  padding: 10px 30px;
  width: 40vw;
`
export const Row = styled(Flex)`
  justify-content: space-between;
  height: 80px;
  align-items: center;
  color: ${colors.lightBlack};
`
export const TextLabel = styled(Flex)`
  width: 20%;
  height: 100%;
  font-size: 20px;
  align-items: center;
  ::after {
    content: '*';
  }
`
export const TextInput = styled.input`
  width: 60%;
  border: 0.5px solid rgba(8, 8, 8, 0.2);
  background: #f3f3f3;
  border-radius: 8px;
  align-items: center;
  padding: 10px 20px;
  font-size: 20px;
`
export const Footer = styled(Flex)`
  margin-top: 10px;
  border-top: 0.1px solid rgba(64, 64, 64, 0.1);
  padding-top: 20px;
  justify-content: flex-end;
`
