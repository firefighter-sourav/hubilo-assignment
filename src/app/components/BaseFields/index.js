import { startCase } from 'lodash'
import PropTypes from 'prop-types'
import React, { useCallback, useContext, useEffect, useState } from 'react'

import { isNullOrEmpty, showToast } from 'app/configs/utils'
import FieldsContext from 'app/contexts/FieldsContext'

import * as Styles from './styles'

import Button from '../Button'

/**
 * BaseFields is a seperate component from BaseInput because of formData
 * formData is state so updating will rerender the component including the modal
 * Updating the modal everytime just for input change is not efficient
 *
 */
const BaseFields = (props) => {
  const { closeModal } = props
  const [formData, setFormData] = useState({})
  const { fields, handleSubmit } = useContext(FieldsContext)

  useEffect(() => {
    let tempFormData = {}
    fields.map((field) => {
      tempFormData[field] = null
    })
    setFormData(tempFormData)
  }, [])

  const handleFormSubmit = useCallback(() => {
    if (isNullOrEmpty(formData)) {
      showToast('All Fields are required', 'warning')
      return
    }
    closeModal()
    handleSubmit(formData)
  }, [formData])

  const handleInputChange = (e) => {
    e.persist()
    setFormData((state) => ({
      ...state,
      [e.target.name]: e.target.value,
    }))
  }

  return (
    <Styles.Wrapper>
      {fields.map((name) => (
        <Styles.Row key={name}>
          <Styles.TextLabel>{startCase(name)}</Styles.TextLabel>
          <Styles.TextInput
            name={name}
            onKeyUp={handleInputChange}
            placeholder="Write Something..."
          />
        </Styles.Row>
      ))}

      <Styles.Footer>
        <Button onClick={handleFormSubmit}>Save</Button>
      </Styles.Footer>
    </Styles.Wrapper>
  )
}
BaseFields.propTypes = {
  closeModal: PropTypes.func.isRequired,
}
BaseFields.defaultProps = {}
export default BaseFields
