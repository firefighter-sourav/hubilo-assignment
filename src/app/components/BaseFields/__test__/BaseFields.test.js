import {
  cleanup,
  render as testRender,
  fireEvent,
} from '@testing-library/react'
import React from 'react'
import ReactDOM from 'react-dom'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'

import BaseFields from '../index.js'

afterEach(cleanup)

it('renders ConfigSelect without crasing', () => {
  const closeModal = jest.fn()
  const div = document.createElement('div')
  ReactDOM.render(<BaseFields closeModal={closeModal} />, div)
  ReactDOM.unmountComponentAtNode(div)
})

it('matches snapshot', () => {
  const closeModal = jest.fn()
  const tree = renderer.create(<BaseFields closeModal={closeModal} />).toJSON()
  expect(tree).toMatchSnapshot()
})
