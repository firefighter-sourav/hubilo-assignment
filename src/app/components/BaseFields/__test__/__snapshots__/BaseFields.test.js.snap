// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-dlnjwi kONwXc"
>
  <div
    className="sc-bdnxRM sc-gKAaRy kGFBfj fIBEnI"
  >
    <div
      className="sc-bdnxRM sc-gtsrHT sc-iCoGMd kGFBfj NYOTq giQWvX"
      data-testid="config-button"
      onClick={[Function]}
    >
      Save
    </div>
  </div>
</div>
`;
